src/integrator/cuda/cuda_inc_potential.cu

```diff
@@ -7,16 +7,25 @@ namespace cuda       {
 
 using namespace GauXC::cuda;
 
-template <typename T>
-__global__ void inc_by_submat_combined_kernel( size_t           ntasks,
-                                               XCTaskDevice<T>* device_tasks,
-                                               T*               A,
-                                               size_t           LDA ) {
 
-  const int batch_id = blockIdx.z;
+#define WARP_X 32
+#define WARP_Y 1
+#define UNROLL_FACTOR 4
+#define EFF_UNROLL 4
+#define CUT_X 8
+#define CUT_Y 4
 
-  if( batch_id < ntasks ) {
 
+template <typename T>
+__global__ __launch_bounds__(1024, 1)
+void inc_by_submat_combined_kernel( size_t           ntasks,
+                                    XCTaskDevice<T>* device_tasks,
+                                    T*               A,
+                                    size_t           LDA, 
+				    const int block_y,
```

Comment:
I noticed that calling in calling this kernel, you make 4 calls (00,01,10,11). Would there be any benefit to just making this a non-type template parameter such that the calling sig is known at compile time? It's unclear to me to what extend `nvcc` exploits `constexpr` type knowledge, but traditional C++ compilers would likely be able to optimize around this knowledge (especially with `if constexpr`)

Reply:
This was hard coded to be 4 calls initially as `A` was only being broken into 4 parts. However, for larger problem sizes, I am breaking the matrix into many different chunks and have a kernel launch for each one.

I am finding that the optimal chunk size varies depending on the problem size and L2 cache of the device, with smaller chunks leading to better cache performance, but also increasing the number of kernel launches and overhead. I am trying to work out some heuristics to choose the optimal size for a given problem. For Taxol the matrix is O(1k) and for Ubiquitin it is O(10k); can we expect most problems to fall within this range, or could there be even larger problems of interest?

Reply:
I believe our stretch targets are O(50k) - O(100k), but I don't currently have examples to that effect. There are also other problems we have to address before we can go that large anyways (i.e. the assumption that the density/VXC can simultaneously fit in device memory will likely not hold). But if we can write up the methodology as to how to select these hyper-parameters, we can address how to handle this generally in the future